src/index.ts

```diff
@@ -0,0 +1,188 @@
+import { parse } from 'graphql';
+import { Location, DocumentNode, DefinitionNode } from 'graphql/language';
+
+// Strip insignificant whitespace
+// Note that this could do a lot more, such as reorder fields etc.
+function normalize(text: string): string {
+  return text.replace(/[\s,]+/g, ' ').trim();
+}
+
+type DocCache = { [docString: string]: DocumentNode };
+let docCache: DocCache = {};
+
+interface FragmentSourceMap {
+  [framentName: string]: {
+    [source: string]: boolean;
+  };
+}
+
+// A map fragmentName -> [normalized source]
+let fragmentSourceMap: FragmentSourceMap = {};
+
+function cacheKeyFromLoc(loc?: Location): string {
+  if (!loc) return '';
+  return normalize(loc.source.body.substring(loc.start, loc.end));
+}
+
+// For testing.
+export function resetCaches(): void {
+  docCache = {};
+  fragmentSourceMap = {};
+}
+
+interface FragmentMap {
+  [key: string]: boolean
+}
+
+// Take a unstripped parsed document (query/mutation or even fragment), and
+// check all fragment definitions, checking for name->source uniqueness.
+// We also want to make sure only unique fragments exist in the document.
+let printFragmentWarnings = true;
+function processFragments(ast: DocumentNode): DocumentNode {
+  const astFragmentMap: FragmentMap = {};
+  const definitions: DefinitionNode[] = [];
+
+  for (let i = 0; i < ast.definitions.length; i++) {
+    const fragmentDefinition = ast.definitions[i];
+
+    if (fragmentDefinition.kind === 'FragmentDefinition') {
+      const fragmentName: string = fragmentDefinition.name.value;
```

Comment:
was this explicitly needed? does `fragmentDefinition.name.value` not expose the `string` type?